docs/mempools/dd.md

```diff
@@ -0,0 +1,106 @@
+- When cleaning up the pool in a loop, loops over pending (executable?) txs for every sender that has gone over AccountSlots, and drops txs (based on nonce) form them. Also loops over future (non-executable?) txs based on **heartbeats**: accounts with the most time without any activity get their txs pruned first.
+
+## Difficulties
+
+In addition to all complications that Ethereum has, we also have the issue that a tx public execution can invalidate an arbitrary number of existing txs just by emitting nullifiers. We have no way of knowing that in advance.
+
+Also, while for Ethereum a "replacement" is just a tx with the same nonce and sender as an existing one, for us any tx that shares a nullifier can technically be a replacement. This also means that tx A may be a replacement for B and C, but B and C may be unrelated to each other.
+
+Also, while our `fee_payer` slightly matches Ethereum's `sender`, it's possible that many users (if not all) use the same very few fee payers (in our case, FPCs), so there is likely no point in setting limits per sender as Ethereum does. Remember that, thanks to privacy, we cannot know the sender of a tx. On the flip side, we know that two txs do come from the same user if they share a private-land nullifier.
+
+## Design
+
+To recap, we need to consider:
+
+- Balance of fee-payers
+- Conflicting nullifiers
+- Max block number
+- Gas fees and limit vs current base fees and limits
+- Archive root (only on reorgs)
+
+We propose keeping the following indices for all txs:
+
+- priority fee
+- fee-payer
+- nullifier
+- base fee
+- gas limit
+- max block number
+
+When adding a tx, in addition to running all static checks:
+
+- We check if the current balance of the fee payer, minus the max cost of all pending txs for that fee payer, is enough to pay for this tx. If it is not, we try evicting other txs with a lower priority fee. If that works, and all other checks pass, we include the tx dropping the others.
+- We check if it shares a nullifier with any existing pending tx. If it pays more than all of the conflicting ones, and it passes all other checks, we include it and drop the other ones.
+- We check if the tx gas limit is below the current block gas limit, and its fees are above the current base fees. If not, we drop it. Note that we could save it for later in case fees drop in the future, but this means tracking two different pools (executable and non-executable, as geth does).
+- We check if we are below a configurable size/number of pending txs. If we are not, start dropping txs with lower priority fee (sorted by priority fee) until we get again below the threshold.
+- If we do add the tx, we index its max block number as the minimum of the tx's max-block-number and the current block number plus a configurable number. This allows us to evict txs after they'd been sitting in the pool for a very long time.
```

Comment:
Maybe i missed it explicitly, but also if the max block number is in the past we just drop it already. The max block number was mentioned earlier for dynamic, but not explicitly discarding etc. Maybe just me being braindead because late.